hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java

```diff
@@ -757,7 +757,7 @@ public boolean next(List<Cell> outResult, ScannerContext scannerContext) throws
         }
       } while ((cell = this.heap.peek()) != null);
 
-      if (count > 0) {
+      if (count > 0 || storeLimit > -1 && this.countPerRow > (storeLimit + storeOffset)) {
```

Comment:
nit: it might look too many brackets but for better readability, good to add an additional bracket after first OR operator `if (a || (b && c))` ?